modules/subschema/src/main/scala/com.snowplowanalytics/iglu.schemaddl/jsonschema/subschema/package.scala

```diff
@@ -80,7 +76,76 @@ package object subschema {
       case _ => s
     }
 
-  def simplify(s: Schema): Schema = s
+  def canonicalizeUnion(s: Schema): Schema = {
+    s.`type` match {
+      case Some(Union(types)) =>
+        val anyOf = types.map(t => Schema.empty.copy(`type` = Some(t))).toList
+        s.copy(`type` = None, anyOf = Some(AnyOf(anyOf)))
+      case _ => s
+    }
+  }
+
+  def inferType(j: Json): (Type, Option[Json]) =
+    // enum values are plain JSON values, not schemas
+    j.fold(
+      jsonNull = Null -> None,
+      jsonBoolean = v => Boolean -> Some(Json.fromBoolean(v)),
+      jsonNumber = v => Number -> Some(Json.fromJsonNumber(v)),
+      jsonString = v => String -> Some(Json.fromString(v)),
+      jsonArray = v => Array -> Some(Json.fromValues(v)),
+      jsonObject = v => Object -> Some(Json.fromJsonObject(v))
+    )
+
+  def simplify(s: Schema): Schema =
+    (simplifyMultiValuedEnum _)
+      .andThen(simplifyEnum)(s)
+
+  def simplifyMultiValuedEnum(s: Schema): Schema =
+    (s.`type`, s.`enum`) match {
+      case (t@Some(typ), Some(Enum(values))) if typ != Boolean && values.length > 1 =>
+        val anyOf = values.map(v => Schema.empty.copy(`type` = t, `enum` = Some(Enum(List(v))))).map(simplifyEnum(_))
+        Schema.empty.copy(anyOf = Some(AnyOf(anyOf)))
+      case _ => s
+    }
+
+  def simplifyEnum(s: Schema): Schema =
+    (s.`type`, s.`enum`) match {
+      case (t@Some(Null), Some(Enum(List(Json.Null)))) =>
+        Schema.empty.copy(`type` = t, `enum` = None)
+      case (t@Some(String), Some(Enum(v :: Nil))) =>
+        Schema.empty.copy(`type` = t, pattern = v.asString.map(s => Pattern(s"^$s$$")))
+      case (t@Some(Integer), Some(Enum(v :: Nil))) =>
+        val intV = v.asNumber.flatMap(_.toBigInt)
+        Schema.empty.copy(`type` = t, minimum = intV.map(IntegerMinimum(_)), maximum = intV.map(IntegerMaximum(_)))
+      case (t@Some(Number), Some(Enum(v :: Nil))) =>
+        val decV = v.asNumber.flatMap(_.toBigDecimal)
+        Schema.empty.copy(`type` = t, minimum = decV.map(NumberMinimum(_)), maximum = decV.map(NumberMaximum(_)))
+      case (t@Some(Array), Some(Enum(vArr :: Nil))) =>
+        val maybeArr = vArr.asArray
+        val arrSize = maybeArr.map(_.size)
+        val tupleItems = maybeArr.map(xv => xv.map(schemaForEnumValue).map(simplifyEnum).toList)
```

Comment:
One thing to note about arrays and objects is that we know the `enum` keyword will contain arrays and objects respectively, but we don't know anything more than that by just looking at the original schema. Since the rule pushes down the `enum` construct to `items` and `properties` respectively, to avoid having a list of untyped enums as a result we infer the type of each single enum in `schemaForEnumValue`, so that we can then recursively invoke `simplifyEnum` and make sure the whole thing is simplified